src/ServiceControl.Persistence.RavenDb5/ErrorMessagesDataStore.cs

```diff
@@ -730,13 +734,37 @@ await using (var ie = await session.Advanced.StreamAsync(query))
                     }
                 }
             }
-
             return ids.ToArray();
         }
 
-        public Task<byte[]> FetchFromFailedMessage(string uniqueMessageId)
+        public async Task<byte[]> FetchFromFailedMessage(string uniqueMessageId)
         {
-            throw new NotSupportedException("Body not stored embedded");
+            byte[] body = null;
+            var result = await bodyStorage.TryFetch(uniqueMessageId);
+
+            if (result == null)
+            {
+                throw new InvalidOperationException("IBodyStorage.TryFetch result cannot be null");
+            }
+
+            if (result.HasResult)
+            {
+                using (result.Stream) // Not strictly required for MemoryStream but might be different behavior in future .NET versions
```

Comment:
Wouldn't it be more efficient to check if Stream is a MemoryStream so that the `CopyToAsync` isn't required?

Reply:
This chunk is [copy-pasted from here](https://github.com/Particular/ServiceControl/blob/18bc698d6a67e4721b39ae6ddcf7c7dddb7592b6/src/ServiceControl/Recoverability/Retrying/Infrastructure/ReturnToSender.cs#L114) so I didn't analyze it to that level.

Reply:
Like we talked about, I'd rather put this in a different PR.